### PR TITLE
CLDR-16239 inheritance; modify -fR post-processed VXML

### PR DESCRIPTION
**Pull request by btangmu**
-CLDRModify.java has new -fR option for Revert

-Revert here means, use VXML but revert to baseline in some cases

-Data in common/main upated with post-processed VXML

CLDR-16239

- [ ] This PR completes the ticket.

ALLOW_MANY_COMMITS=true

### CODE REVIEW

**Comment by btangmu** [comment]
Observations:

- many values change to ^^^, surprising since this is based on "drop=false" vxml
- problem remains: af "{0} swaartekrag" changed to "{0} G", no reversion to baseline since vxmlValue is not ^^^

**Comment by btangmu** [comment]
work in progress

**Comment by srl295** [comment]
@btangmu did you mean to merge this into the https://github.com/unicode-org/cldr/tree/brs/v43vxml branch instead. You can just retarget it by clicking Edit on the PR

**Comment by btangmu** [comment]
@srl295 I did retarget to [brs/v43vxml](https://github.com/unicode-org/cldr/tree/brs/v43vxml) as you suggested, thanks!

File: tools/cldr-code/src/main/java/org/unicode/cldr/tool/CLDRModify.java [hunk 1]
```diff
@@ -2203,6 +2186,126 @@ public void handlePath(String xpath) {
...
+            }
+
+            @Override
+            public void handlePath(String xpath) {
+                if (vxmlFile == null) {
+                    return; // use baseline
+                }
+                String vxmlValue = vxmlFile.getStringValue(xpath);
+                // steps >= 2 identifies "L2+" locales
+                boolean revertToBaseline = (steps >= 2) && wantRevertToBaseline(xpath, vxmlValue);
```
Comment (btangmu) [inline tools/cldr-code/src/main/java/org/unicode/cldr/tool/CLDRModify.java:2231]: possibly drop the L2+

File: tools/cldr-code/src/main/java/org/unicode/cldr/tool/CLDRModify.java [hunk 2]
```diff
@@ -2203,6 +2186,126 @@ public void handlePath(String xpath) {
...
+                    // criterion 3: if bailey value is not from different path and locale, don't revert to baseline
+                    return false;
+                }
+                return true;
+            }
+
+            private boolean changesWereAllowed(String localeID, String xpath) {
+                boolean isError = pathHasError(localeID, xpath);
+                String oldValue = baselineFileUnresolved.getWinningValue(xpath);
+                boolean isMissing = (oldValue == null);
```
Comment (btangmu) [inline tools/cldr-code/src/main/java/org/unicode/cldr/tool/CLDRModify.java:2275]: if old value = null or if path.contains @draft=

File: tools/cldr-code/src/main/java/org/unicode/cldr/tool/CLDRModify.java [hunk 3]
```diff
@@ -2203,6 +2186,126 @@ public void handlePath(String xpath) {
...
+             * 	Revert each L2+ locale/path VXML value of ↑↑↑ to the baseline value, where all the following criteria are met:
+             * 	1	The baseline value is a hard value (can use unresolved file for this)
+             * 	2	The survey tool did not allow changes in the locale/path in v43
+             * 	    ▪ That is, other than allowed locales with Person Names, Turkey, missing, and errors
+             *    	▪ NOTE: we should be able to check missing/errors by running the CheckCLDR tests on the baseline data. Most locales only had PN & Turkey, but some had errors or missing values.
+             * 	3	The bailey value is fetched from a different path AND from an ancestor locale.
+             * 	    ▪ Ideally we would look at intervening paths also, but I think this is sufficient.
+             */
+            private boolean wantRevertToBaseline(String xpath, String vxmlValue) {
+                if (!CldrUtility.INHERITANCE_MARKER.equals(vxmlValue)) {
```
Comment (btangmu) [inline tools/cldr-code/src/main/java/org/unicode/cldr/tool/CLDRModify.java:2248]: could vxmlValue be null?

File: tools/cldr-code/src/main/java/org/unicode/cldr/tool/CLDRModify.java (hunk 1 above)
Comment (btangmu) [reply tools/cldr-code/src/main/java/org/unicode/cldr/tool/CLDRModify.java:2231]: and test if vxmlValue is null -- throw an exception!

File: tools/cldr-code/src/main/java/org/unicode/cldr/tool/CLDRModify.java [hunk 4]
```diff
@@ -2203,6 +2186,126 @@ public void handlePath(String xpath) {
...
+            @Override
+            public void handlePath(String xpath) {
+                if (vxmlFile == null) {
+                    return; // use baseline
+                }
+                String vxmlValue = vxmlFile.getStringValue(xpath);
+                // steps >= 2 identifies "L2+" locales
+                boolean revertToBaseline = (steps >= 2) && wantRevertToBaseline(xpath, vxmlValue);
+                if (!revertToBaseline) {
+                    String fullXPath = cldrFileToFilter.getFullXPath(xpath);
```
Comment (btangmu) [inline tools/cldr-code/src/main/java/org/unicode/cldr/tool/CLDRModify.java:2233]: use fullpath from the vxml

File: tools/cldr-code/src/main/java/org/unicode/cldr/tool/CLDRModify.java (hunk 2 above)
Comment (btangmu) [reply tools/cldr-code/src/main/java/org/unicode/cldr/tool/CLDRModify.java:2275]: needs to be the full-vxml-path (extra arg)

Comment (btangmu) [reply tools/cldr-code/src/main/java/org/unicode/cldr/tool/CLDRModify.java:2275]: could pass in the draft status of the full-vxml-xpath as arg to this method

Comment (srl295) [reply tools/cldr-code/src/main/java/org/unicode/cldr/tool/CLDRModify.java:2275]: `CLDRFile.DraftStatus.forXPath(fullXpath) == CLDRFile.DraftStatus.provisional`

File: tools/cldr-code/src/main/java/org/unicode/cldr/tool/CLDRModify.java [hunk 5]
```diff
@@ -2203,6 +2186,126 @@ public void handlePath(String xpath) {
...
+            @Override
+            public void handleStart() {
+                if (vxmlFactory == null) {
+                    vxmlFactory = SimpleFactory.make(list, ".*");
+                    if (!pathHasError("zh_Hant", "//ldml/personNames/sampleName[@item=\"foreignFull\"]/nameField[@type=\"surname-core\"]")) {
+                        throw new RuntimeException("pathHasError wrong?");
+                    }
+                }
+                String localeID = cldrFileToFilter.getLocaleID();
+                baselineFileUnresolved = cldrFileToFilter.getUnresolved();
```
Comment (btangmu) [inline tools/cldr-code/src/main/java/org/unicode/cldr/tool/CLDRModify.java:2214]: cldrFileToFilter is maybe already unresolved? (unless -z on command line?)

Comment (btangmu) [reply tools/cldr-code/src/main/java/org/unicode/cldr/tool/CLDRModify.java:2214]: try with getResolved...

File: tools/cldr-code/src/main/java/org/unicode/cldr/tool/CLDRModify.java (hunk 2 above)
Comment (srl295) [reply tools/cldr-code/src/main/java/org/unicode/cldr/tool/CLDRModify.java:2275]: @btangmu you might want to add this to CLDRFile.DraftStatus:

```java
/**
 * provisional.orWorse(x) - true for unconfirmed, provisional…  false for contributed, approved
 */
public boolean orWorse(DraftStatus d) {
    return d.ordinal() <= this.ordinal();
}
```

Comment (srl295) [reply tools/cldr-code/src/main/java/org/unicode/cldr/tool/CLDRModify.java:2275]: then you could do:

`if ( CLDRFile.DraftStatus.provisional.orWorse(CLDRFile.DraftStatus.forXPath(fullXPath)))`

true if provisional or worse

File: common/main/af.xml [hunk 6]
```diff
@@ -6160,7 +6160,7 @@ Warnings: All cp values have U+FE0F characters removed. See /annotationsDerived/
 			</compoundUnit>
 			<unit type="acceleration-g-force">
 				<displayName>swaartekrag</displayName>
-				<unitPattern count="one">{0} swaartekrag</unitPattern>
+				<unitPattern count="one">↑↑↑</unitPattern>
```
Comment (btangmu) [inline common/main/af.xml:6163]: what makes this clearly wrong is it's replacing a hard value with an inheritance marker

and in a path that didn't allow changes

Comment (btangmu) [reply common/main/af.xml:6163]: "clearly" being subjective and not the adverb I would use - the above note is from what was said in the meeting

File: tools/cldr-code/src/main/java/org/unicode/cldr/tool/CLDRModify.java (hunk 2 above)
Comment (btangmu) [reply tools/cldr-code/src/main/java/org/unicode/cldr/tool/CLDRModify.java:2275]: I'd rather avoid changes outside of CLDRModify until the dust has settled, so for the next PR I'll do something equivalent without changing CLDRFile:

`boolean isMissing = (oldValue == null || CLDRFile.DraftStatus.forXpath(fullXPath).ordinal() <= CLDRFile.DraftStatus.provisional.ordinal());`

Let's remember the idea for orWorse().

Comment (macchiati) [reply tools/cldr-code/src/main/java/org/unicode/cldr/tool/CLDRModify.java:2275]: That's fine for now.

BTW, later, better would be 

```
        public boolean isWorseThan(DraftStatus target) {
            return compareTo(target) < 0;
        }

```
Usage would be 

`DraftStatus.forXpath(fullXPath).isWorseThan(DraftStatus.contributed)`

**Comment by srl295** [comment]
@btangmu if this is obsolete may want to close.

**Pull request closed without merging**